### ISSUE TITLE
How can I deal with the problem - "This model's maximum context length is 4097 tokens. "?

### ISSUE THREAD

**Issue by popyoung** [OkGoDoIt/OpenAI-API-dotnet#95]
I had a conversation with chatGPT, and after several rounds it came with this error. Is there anyway to continue this conversation?

Error at chat/completions (https://api.openai.com/v1/chat/completions) with HTTP status code: BadRequest. Content: {
  "error": {
    "message": "This model's maximum context length is 4097 tokens. However, your messages resulted in 4123 tokens. Please reduce the length of the messages.",
    "type": "invalid_request_error",
    "param": "messages",
    "code": "context_length_exceeded"
  }
}

**Comment by popyoung** [comment]
As a side note, I would like to say that I use these two functions: 
```
Chat.AppendUserInput(question);
string response = await Chat.GetResponseFromChatbot();
```
I took a look at the code of them. It seems that every time I call AppendUserInput, it adds a message to the list.
And eventually list of messages will reach the limit.

**Comment by frankcarey** [comment]
I don't know of a way that automates this.

You'll have to track how many tokens your messages are using total (plus the max amount you want to generate) and then once you reach the limit for a particular model, you'll have to cut back somewhere. I've used a summarization call to take a chunk of responses and cut them down to a single summary with fewer characters, or you can simply start to "forget" earlier messages by deleting the first N messages.

**Comment by popyoung** [comment]
Indeed, I have found it is inevitable. Is there a good way to count the token number?

**Comment by j4587698** [comment]
May be you can use https://github.com/aiqinxuancai/TiktokenSharp .
It can do it

**Comment by Tronald** [comment]
Anyone ever find a good solution to this? Appending the chat is great until it grows too large, but there appears to be no way to know when that point hits, nor can you remove history from the appended messages. A FIFO approach once the token limit is exceeded, that begins stripping the First In items would be great in theory, just not sure how to execute via an existing `Conversation`.

**Comment by OkGoDoIt** [comment]
I addressed this in [version 1.8](https://github.com/OkGoDoIt/OpenAI-API-dotnet/releases/tag/v1.8) released December 6th.

Details at https://github.com/OkGoDoIt/OpenAI-API-dotnet#Conversation-History-Context-Length-Management

Basically:

1. If the context length is too long, it will automatically remove the oldest non-system messages from the chat history and try to resubmit. 
2. You can override this behavior by implementing `chat.OnTruncationNeeded` or setting `chat.AutoTruncateOnContextLengthExceeded = false`
3. You can also now manually edit the chat message history list, it is no longer readonly
